Refactoring, premier état

laowantong opened the pull request:
J'ai fait une première passe de refactoring jusqu'au lancement de `run_fptas`. Par contre, ça donne des Cmax nuls, donc j'imagine que j'ai loupé un truc :

```
$ python run_fptas.py 
Beginning of the run at 2020-01-23 21:05:00.958624
input number : input/logs/H3-nbT5-006.json at 2020-01-23 21:05:00.959160
Time needed for FPTAS:  9.918212890625e-05
FPTAS : Cmax = 0; 60 states generated
Time needed for FPTAS:  8.20159912109375e-05
FPTAS : Cmax = 0; 60 states generated
input number : input/logs/H3-nbT5-001.json at 2020-01-23 21:05:00.959823
Time needed for FPTAS:  6.699562072753906e-05
...
```

Dommage qu'il n'y ait pas de tests pour cette partie, c'est là qu'ils auraient été utiles !

Peux-tu jeter un coup d'œil ?

NB: Je n'ai pas essayé de comprendre ou de reprendre de larges parties de ton code dont l'utilité ne m'apparaissait pas franchement. Il y a manifestement beaucoup de code mort ou inutilisé. Dans la présente branche, pour référence, tu trouveras tes anciens programmes sous `deprecated`.

laowantong committed:
making the fptas solver into a class (example)

laowantong committed:
work in progress

laowantong committed:
Revert "work in progress"

This reverts commit 62cd3e60848cd5af80771fc82c108d9c175c7f16.

laowantong committed:
add improved fptas, first take

laowantong committed:
Merge branch 'master' into objects

laowantong committed:
backporting #9

laowantong committed:
backport deduplication

laowantong committed:
make FPTAS solver into a class

SarahMinich committed:
Updated the gitignore file

laowantong committed:
Create normalize_json.py

laowantong committed:
Merge branch 'normalize_json' into refactoring

laowantong committed:
add symbol, tile, instance and their tests

laowantong committed:
Merge branch 'objects' into refactoring

laowantong committed:
suppress encoding shebang

laowantong committed:
update badges after migration under pagination-problem

laowantong committed:
fix #1

laowantong committed:
deprecate legacy code

laowantong committed:
rename old test files to prevent Travis CI to launch them and fail the build

laowantong committed:
amend last commit

laowantong committed:
add Page class and its tests

laowantong committed:
abbreviate "directory" in "dir" in config files

laowantong committed:
add a minimal Instance class

laowantong committed:
make fptas run again

laowantong committed:
dedent main program

SarahMinich commented:
J'ai exécuté le FPTAS à la main sur l'instance H3-nbT5-001, voici en photo ce que j'obtiens.

![H3-nbT5-001_execution_manuelle_1](https://user-images.githubusercontent.com/53268501/73535144-285dc700-4423-11ea-9a7e-18995cd4e893.JPG)
![H3-nbT5-001_execution_manuelle_2](https://user-images.githubusercontent.com/53268501/73535146-285dc700-4423-11ea-8d75-a413512c1817.JPG)

Je me doute que sur papier, ce n'est pas pratique mais je ne sais pas quel format serait le plus facile d'utilisation pour vous. Dites moi comment vous voulez que je le tape et je le ferai !

Je ne sais pas non plus pourquoi les photos s'affichent tournées dans le mauvais sens en prévisualisation mais si on clique dessus, elles s'affichent dans le bon sens !

SarahMinich committed:
Added a paper version of a test for the FPTAS on input H3-nbT5-001

laowantong committed:
Revert "Added a paper version of a test for the FPTAS on input H3-nbT5-001"

This reverts commit f181a8fa2f9c2005d0508fa16aeac5c4394a5c8a.

laowantong commented:
J'ai annulé ton dernier commit. GitHub n'est pas Dropbox. Les photos n'ont rien à faire dedans. Dans un commentaire d'issue, c'est ok bien sûr.

laowantong commented:
Les tests manuels ne peuvent pas être automatisés, donc ça n'aide pas du tout. Il faudrait que ton solveur soit pourvu de tests unitaires, de sorte qu'on puisse tester ses fonctions une par une. Ne fais aucun commit dans cette branche, s'il te plaît. Mais ajoute des tests pour ta version du programme qui est dans le master. Si tu dois modifier le programme lui-même pour pouvoir le tester, fais-le dans une branche à toi.

SarahMinich commented:
Je ne sais pas quoi faire d'autres que comparer les résultats obtenus par les traitements faits par les sous-programmes du fichier fptas.py avec ceux qu'on est supposés obtenir sur une instance pour vérifier si les sous-programmes fonctionnent correctement donc si c'est ce que vous appelez des tests manuels, je ne sais pas quoi faire d'autre.

laowantong commented:
> comparer les résultats obtenus par les traitements faits par les **sous-programmes** du fichier fptas.py avec ceux qu'on est supposés obtenir sur une instance pour vérifier si les sous-programmes fonctionnent correctement

Il faudrait tester ça automatiquement, oui.

> ce que vous appelez des tests manuels

= faits à la main sur une feuille avec un crayon

laowantong committed:
fix c_max==0 problem

laowantong commented:
@SarahMinich 

Bon, tous les tests passent maintenant, il suffisait de transmettre `chi_seed`, ce qui m'avait échappé en découplant les solveurs `"basic"` et `"improved"` (lignes 45, 58 et 76 ci-dessous).

https://github.com/pagination-problem/tree_structure/blob/8e53f36220a485f6512e9a0b21b2c75823124a25/fptas.py#L45-L76

Les scripts et les tests de cette branche sont normalement fonctionnellement équivalents aux anciens. C'est maintenant à toi de jouer, à travers les étapes suivantes :

1. Lire attentivement le nouveau code et ses tests, en le comparant aux anciennes versions (par exemple, switche sur la branche `refactoring` sous VS-code, et regarde le `master` sous ton navigateur). Bien sûr tu dois te familiariser avec les nouvelles interfaces des fonctions / méthodes / classes et les nouveaux formats de données, puisque c'est ça que tu devras utiliser dans la suite. Mais je suis sûr que tu peux aussi apprendre beaucoup de choses au passage, qui ne t'ont pas forcément été enseignées lors de ta formation (c'est un autre problème).
2. Déplacer le code obsolète de la branche `master` dans le dossier `deprecated` prévu à ce effet. **Attention !** tu dois le faire en ligne de commande pour que GitHub s'y retrouve :
```
git mv old_file deprecated
```
3. S'il reste du code « vivant » et que tu veux garder :
    1. écrire les tests correspondants (dans `master`) ;
    2. attendre que j'aie revu et corrigé ton code et tes tests, ce que je continuerai à faire dans la présente branche ;
    3. GOTO 1.
4. Une fois qu'on aura une base de code saine, on mergera `refactoring` dans `master` et je créerai une autre branche pour faire le même travail de simplification, mais cette fois sur les structures de données. Avec, parmi les objectifs, l'utilisation de bibliothèques rapides comme Numpy ou de techniques d'accélération comme Numba.

laowantong committed:
homogeneize types in array (replace 'x' by NA == -1)

SarahMinich commented:
Je pense avoir compris les modifications faites au code, il ne me reste que quelques interrogations.

1. Mon plus gros problème est la classe Symbol, je ne comprends pas du tout comment elle fonctionne. Comment et quand puis-je initialiser la taille d'un symbole ?
2. Dans fptas.py à la ligne 35, si on appelle S1 (resp. S2) l'ensemble des symboles de t1 (resp. t2) quand on fait
`t1.symbols.difference(t2.symbols)`. 
Cette instruction est bien équivalente à l'opération ensembliste S1 \ S2 ? 
3. J'aimerais garder le grand commentaire du début que j'avais mis dans fptas.py de la branche master, pour me rappeler la "gymnastique" nécessaire dans les numéros des feuilles (celui avec l'arbre "dessiné").
4. Dans instance.py : je crois qu'on ne peut plus créer d'instance autrement qu'en en lisant une depuis un fichier d'instance (ce qui est un problème je pense). Inversement je ne suis pas sûre de savoir comment créer un fichier json à partir d'un objet de type Instance (est-ce que j'appelle la fonction ' deserialize ' puis je fais l'équivalent d'un write-json du data obtenu ?
5. page.py n'est pas nécessaire mais je ne sais pas où écrire la ligne de commande dont vous parlez plus haut ?

laowantong commented:
1. Cf. `test_symbol.py`.
2. Oui, cf. https://docs.python.org/3.8/library/stdtypes.html#set-types-set-frozenset.
3. Tu peux bien sûr réinjecter les commentaires que tu veux, en gardant à l'esprit qu'un commentaire n'est qu'un pis-aller pour ce qu'on ne peut exprimer clairement ni dans le code, ni dans les tests : typiquement, ils servent à expliquer _pourquoi_ on fait (ou on ne fait pas) telle chose. Ce que tu sembles vouloir écrire en l'occurrence est une chaîne de documentation (Google "docstring").
4. Effectivement, si c'est un besoin réel, on peut facilement rajouter la création d'une instance à partir d'un dictionnaire ou d'une chaîne JSON. Inversement, on peut rajouter une méthode `serialize()` qui renvoie une chaîne JSON, ou `dump()` qui l'écrit directement. Ça peut se faire en reprenant la technique que j'ai utilisée dans `normalize_json.py` (lignes 22-24).
5. Sous le terminal de VS-code par exemple.

SarahMinich committed:
Added a docstring for the fptas.py file

SarahMinich committed:
Moved to deprecated the page.py file

SarahMinich commented:
1. J'ai regardé test_symbol.py, j'ai mal formulé ma question. Je sais faire la ligne d'instruction puisque c'est la même que pour une classe "classique" mais je ne comprend pas comment ça peut marcher puisqu'il n'y a pas de fonction init dans la classe Symbol.

2. J'ai trouvé un [lien ](https://www.geeksforgeeks.org/python-set-difference/)qui est (pour moi) plus facile à comprendre car ils ont expliqué ça avec des diagrammes de Venn ! 

3. J'ai regardé et j'ai fait une tentative !

4. D'accord, je regarderai en détails demain !

5. Normalement, c'est fait !

laowantong committed:
modify and extend Instance class interface (cf. #2)

Methods:
- remove .deserialize()
- add .get_data()
- add .get_json()
- add.dump_json()

An instance can now be intialized with either a path towards a JSON file, or a dictionary whose structure is the same as obtained via a json.loads() deserialization.

laowantong committed:
Merge branch 'refactoring' of https://github.com/pagination-problem/tree_structure into refactoring

laowantong committed:
prevent warning on invalid escape sequence \

laowantong committed:
restore page.py

laowantong commented:
1. Cherche `class typing.NamedTuple` sur https://docs.python.org/3/library/typing.html.
2. Ok.
3. En introduisant un warning malheureusement, cf. 228543c.
4. Je l'ai fait en f41ed74.
5. J'ai restauré ce fichier à son emplacement (5ea6b73). Ça me paraît un objet fondamental dans le problème, je ne vois pas pourquoi on ne s'en servirait pas, maintenant ou plus tard. En tout cas il est testé, le déplacer fait donc échouer les tests.

Tu devrais acquérir le réflexe de lancer les tests avant chaque commit (cf. 3 et 5 ci-dessus).

Par ailleurs, je t'ai demandé de ne pas faire de modifications dans cette branche sur laquelle je travaille, précisément pour éviter que tu casses quoi que ce soit. J'ai donné des instructions précises sur les étapes à suivre, je sais que c'était il y a déjà 18 jours, merci de relire si tu as oublié.

laowantong committed:
fix #4

laowantong commented:
OK pour fusionner cette branche dans le master ?

SarahMinich commented:
Je pense que tout est ok pour qu'on fusionne oui !

laowantong committed:
create instance generator (generic types only)

laowantong committed:
make possible tiles of unequal lengths

laowantong committed:
release candidate for random instance generation (cf. #6)

laowantong committed:
simplify main loop

laowantong committed:
test arity > 2

laowantong committed:
document and test json_wrapper()

laowantong committed:
copy instance name into its own data

laowantong committed:
improve clarity

laowantong committed:
remove duplicated code

laowantong committed:
complete previous commit

laowantong committed:
add some comments to matrix rows in test_fptas.py

laowantong committed:
avoid useless calculations

laowantong commented:
Je vais continuer à travailler dans cette branche avant de la fusionner.
